Add proper underline and strikethrough support

chrisduerr opened the pull request:
Support for strikethrough has been added by inserting and removing a
`STRIKE_THROUGH` flag on the cell.

Now all strikethrough and underline drawing is also done through the
rectangle renderer. So no glyphs are used to render underlines and
strikethrough.
The position is taken from the font metrics and should be accurate for
linux, however is not yet tested on macos.

It works by checking the underline state for each cell and then drawing
from the start until the last position whenever an underline ended. This
adds a few checks even if no underline is rendered but I was not able to
measure any significant performance impact.

Fixes jwilm/alacritty#806.
Fixes jwilm/alacritty#31.

Demo:
![](https://u.teknik.io/Nwzo3.png)

TODO:
 - [x] Account for font and glyph offsets
 - [x] Account for window padding
 - [x] Test (and potentially fix) DPI
 - [x] Refactor code and extract it into separate files
 - [x] Add Windows support

chrisduerr commented:
Here are the results of the benchmarks I've run:

<details>
<summary>Master</summary>

```
Benchmark #1: alt-screen-random-write
  Time (mean ± σ):      5.433 s ±  0.032 s    [User: 5.200 s, System: 1.888 s]
  Range (min … max):    5.381 s …  5.476 s
Benchmark #1: scrolling
  Time (mean ± σ):      5.562 s ±  0.087 s    [User: 5.419 s, System: 5.623 s]
  Range (min … max):    5.415 s …  5.714 s
Benchmark #1: scrolling-in-region
  Time (mean ± σ):      5.533 s ±  0.062 s    [User: 5.359 s, System: 5.638 s]
  Range (min … max):    5.399 s …  5.615 s
Benchmark #1: startup
  Time (mean ± σ):      85.7 ms ±   2.2 ms    [User: 73.7 ms, System: 15.0 ms]
  Range (min … max):    83.0 ms …  91.0 ms
Benchmark #1: unicode-random-write
  Time (mean ± σ):      5.494 s ±  0.649 s    [User: 5.469 s, System: 0.155 s]
  Range (min … max):    4.476 s …  6.671 s
```

</details>

<details>
<summary>Underline + Strikethrough</summary>

```
Benchmark #1: alt-screen-random-write
  Time (mean ± σ):      5.409 s ±  0.015 s    [User: 5.164 s, System: 1.930 s]
  Range (min … max):    5.382 s …  5.431 s
Benchmark #1: underline-strikethrough-out/scrolling
  Time (mean ± σ):      5.516 s ±  0.064 s    [User: 5.350 s, System: 5.623 s]
  Range (min … max):    5.432 s …  5.615 s
Benchmark #1: scrolling-in-region
  Time (mean ± σ):      5.528 s ±  0.072 s    [User: 5.364 s, System: 5.614 s]
  Range (min … max):    5.432 s …  5.665 s
Benchmark #1: startup
  Time (mean ± σ):      85.3 ms ±   1.9 ms    [User: 72.8 ms, System: 16.0 ms]
  Range (min … max):    82.0 ms …  89.6 ms
Benchmark #1: unicode-random-write
  Time (mean ± σ):      5.836 s ±  0.454 s    [User: 5.804 s, System: 0.175 s]
  Range (min … max):    5.032 s …  6.367 s
```

</details>

<details>
<summary>Benchmark Script</summary>

```
#!/bin/bash

# Read the command line arguments
if [[ "$#" -lt 2 ]]; then
    echo "Usage: bench.sh <TERM> <OUT_DIR>"
    exit 1
fi
term="$1"
dir="$2"
echo "Benchmarking '$1' and writing results to '$2'"
echo ""

# Install hyperfine benchmarker
echo "Searching for hyperfine"
if hash hyperfine 2>/dev/null; then
    echo "Found existing hyperfine install"
else
    echo "Installing hyperfine"
    cargo install hyperfine
fi
echo ""

# Create directory for the benchmarks
if [ -d "$dir" ] || [ -f "$dir" ]; then
    echo "Removing existing output directory"
    rm -rf "$dir"
fi
echo "Creating new output directory"
mkdir "$dir"
echo ""

# Generate the benchmarks, aiming for ~5 secs
echo "Generating benchmarks"
echo "Generating alt-screen-random-write benchmark"
cargo run --release -- -w $(tput cols) -h $(tput lines) -sb 599999999 alt-screen-random-write > "$dir/alt-screen-random-write.vte"
echo "Generating scrolling benchmark"
cargo run --release -- -w $(tput cols) -h $(tput lines) -sb 50000000 scrolling > "$dir/scrolling.vte"
echo "Generating scrolling-in-region benchmark"
cargo run --release -- -w $(tput cols) -h $(tput lines) -sb 50000000 scrolling-in-region > "$dir/scrolling-in-region.vte"
echo "Generating unicode-random-write benchmark"
cargo run --release -- -w $(tput cols) -h $(tput lines) -sb 800000 unicode-random-write > "$dir/unicode-random-write.vte"
echo ""

# Run the benchmarks
echo "Running benchmarks"
echo "Running startup benchmark"
hyperfine --style 'basic' "$term -e true" > "$dir/startup.bench"
echo "Running alt-screen-random-write benchmark"
hyperfine --style 'basic' "$term -e cat $dir/alt-screen-random-write.vte" > "$dir/alt-screen-random-write.bench"
echo "Running scrolling benchmark"
hyperfine --style 'basic' "$term -e cat $dir/scrolling.vte" > "$dir/scrolling.bench"
echo "Running scrolling-in-region benchmark"
hyperfine --style 'basic' "$term -e cat $dir/scrolling-in-region.vte" > "$dir/scrolling-in-region.bench"
echo "Running unicode-random-write benchmark"
hyperfine --style 'basic' "$term -e cat $dir/unicode-random-write.vte" > "$dir/unicode-random-write.bench"
echo ""

echo "All benchmarks completed successfully"
echo "You can find the results in $dir/*.bench"
```

</details>

maximbaz commented:
As usual for your PRs, super awesome!

I tested my font (Consolas) as well as another one that I like (Source Code Pro), I have two comments/questions:

1. Underline seems to "jump vertically" depending on a font, wouldn't it be visually nicer to have it exactly in line with the bottom part of block cursor? Have a look at "Consolas - PR", the underline is so much away from the text that it even feels closer to the letters "ikth" of the word "strikethrough" below.

1. The strikethrough like feels too high in "Source Code Pro - PR" example, especially comparing to "Consolas - PR" or the image you showed in your demo.

![image](https://user-images.githubusercontent.com/1177900/35707675-00498b48-07ab-11e8-8bbc-0958feb59c4a.png)

IMO LibreOffice renders both fonts more beautifully, underline is closer to the text in Consolas and strikethrough is centered in Source Code Pro.

![image](https://user-images.githubusercontent.com/1177900/35707689-0d732112-07ab-11e8-8b42-413bae4e704c.png)

Closer comparison:

![image](https://user-images.githubusercontent.com/1177900/35707800-7fc929e6-07ab-11e8-9bc5-76a513b639ca.png)

![image](https://user-images.githubusercontent.com/1177900/35707853-bb67392a-07ab-11e8-8a6b-f801275a4cba.png)

chrisduerr commented:
For the underline I'm taking the metrics that are provided by the font itself, so those should be okay in theory.

However with the strikethrough I'm just taking the cell height and applying strike-through half way through the cell. It might be a better approach to take the font baseline and the ascent and drawing the strikethrough half way through there.

Thanks for the quick feedback. :)

chrisduerr commented:
According to the ![docs](https://www.freetype.org/freetype2/docs/tutorial/step2.html):
`When displaying or rendering underlined text, this value corresponds to the vertical position, relative to the baseline, of the underline bar's center.`

So I'm currently still doing it wrong because I'm positioning the underline relative to the baseline, of the underline bar's top. Not `relative to the baseline, of the underline bar's center`.
This should help moving the underline in the Consolas example from @maximbaz a bit up.

I can't find any specifics on strikethrough though.

chrisduerr commented:
After looking a bit more into it, it seems like there is no "standard" way for rendering the strikethrough. So where exactly the strikethrough line is placed seems to vary significantly from terminal to terminal. From the ones I've looked at none seems perfect.

Another thing I've noticed is that the underline position and thickness are always the same, no matter what font size. I'm not entirely sure if that's correct, because fonts with a huge underline offset (Hack has -4) will lead to the underline leaving the glyph box at small sizes.

maximbaz commented:
Huh, interesting discovery about the underline position - to illustrate:

![image](https://user-images.githubusercontent.com/1177900/35758388-175bf572-0874-11e8-9278-a96579eea89e.png)

maximbaz commented:
The underline raised a little higher with the latest commit, do you have some ideas about how to fix the underline being so low when the font size is very small? I'm just curious, given the benefits and that it works for usual font sizes I personally think it's fine to merge this even in the current state.

This is font size of `4`, Consolas:

![image](https://user-images.githubusercontent.com/1177900/35819206-501dc1bc-0aa2-11e8-802e-b651281eeb09.png)

`10` (my regular font size):

![image](https://user-images.githubusercontent.com/1177900/35819398-e6fd0354-0aa2-11e8-8b16-4c7e7a25d557.png)

`12` (too large for me, but look how the position of underline perfectly aligns with the cursor):

![image](https://user-images.githubusercontent.com/1177900/35819436-02ef20b0-0aa3-11e8-8208-10c8710a90f8.png)

chrisduerr commented:
I'll be honest here, the state this PR is in right now is pretty much completely unusable with some font/size combinations. And I have no idea how exactly freetype underline and position is supposed to work.

I'll have to look more into drawing underline with freetype, maybe check out how some other applications do it. But as long as underline position/width is always the same with all font sizes, this will not work properly.

chrisduerr commented:
Apparently the undreline position and thickness are specified in font units, not device pixels, so they need to be converted first.

There also is an exact specification for strikethrough position and thickness using the `freetype::tt_os2::TrueTypeOS2Table` struct. This gives us access to the `y_strikeout_size` and `y_strikeout_position` methods.

I'll update this PR with a reworked version of this as soon as possible, so others can test it with their font setup.

chrisduerr commented:
Underline has been reworked, however there's still one small caveat:

```
To make sure the underline doesn't disappear with small fonts, the
minimum underline thickness has been set to `1`. This decision was made
because the `Hack` font has an underline thickness under `0.5` with
relatively big fonts.

This edge-case can lead to some odd behavior when making the font size
super tiny (smaller than readable) so just not rendering any underline
might be a viable alternative.
```

I'd love some feedback on the issue if underline should always be rendered.

Strikethrough will hopefully be up tomorrow. :)

maximbaz commented:
Very cool, now the position of the underline stays relatively on the same place regardless of the font size. 

Also, a nice accident, using Consolas with my regular font size of 10 the underline perfectly matches the bottom of the box cursor, just like I wanted 🙂 This isn't necessarily true for all fonts, but it's a pleasant surprise.

In fact, the perfect match is for font sizes from 1 to 12, and starting from font size 13 it goes out of sync by one pixel... but it doesn't matter.

![underline](https://user-images.githubusercontent.com/1177900/36288002-dea3db9c-12b8-11e8-926a-922e9ec49642.gif)

And this is Source Code Pro:

![underline-source-code-pro](https://user-images.githubusercontent.com/1177900/36288192-d2cda54a-12b9-11e8-9ea4-ac9924ec1424.gif)

chrisduerr commented:
I'd classify the first GIF as a bug. It might be up to freetype standards (not sure, I'd have to look into it), but the underline should never be even a single pixel below the cell height.

Since the cell height is the full height of the line, any pixel below that is part of the next line and should never contain content from the previous one. It's probably just a rounding/edge case thing, thanks for always being so quick to test things. :)

chrisduerr commented:
The current breakage on macos is intentional. I will need some time to test this with a VM.

chrisduerr commented:
@maximbaz To check if the font specifies an incorrect underline position/thickness or if there's some kind of off-by-one error, it would really help if you would give me the metrics for the situation where the underline goes one pixel (or more) beyond the block cursor's bottom line.

Here's a patch you can apply to this PR (a0270d9) for printing the metrics to stdout:
https://hastebin.com/kafalerami

maximbaz commented:
Happy to help 🙂

The underline is pixel-perfect on font sizes [5-12, 18, 21] and is one pixel off on font sizes [13-17, 19-20].

<details>
<summary>Click</summary>

```
5
-----------
CELL_HEIGHT: 17
DESCENT: -4
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 0.45117188
CALC POS: -3
CALC THI: 1
-----------

6
-----------
CELL_HEIGHT: 20
DESCENT: -5
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 0.5415039
CALC POS: -4
CALC THI: 1
-----------

7
-----------
CELL_HEIGHT: 24
DESCENT: -6
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 0.63183594
CALC POS: -5
CALC THI: 1
-----------

8
-----------
CELL_HEIGHT: 27
DESCENT: -6
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 0.72216797
CALC POS: -5
CALC THI: 2
-----------

9
-----------
CELL_HEIGHT: 30
DESCENT: -7
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 0.8125
CALC POS: -6
CALC THI: 2
-----------

10
-----------
CELL_HEIGHT: 34
DESCENT: -8
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 0.90234375
CALC POS: -7
CALC THI: 2
-----------

11
-----------
CELL_HEIGHT: 37
DESCENT: -8
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 0.9926758
CALC POS: -7
CALC THI: 2
-----------

12
-----------
CELL_HEIGHT: 41
DESCENT: -9
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 1.0830078
CALC POS: -8
CALC THI: 2
-----------

13
-----------
CELL_HEIGHT: 44
DESCENT: -10
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 1.1733398
CALC POS: -9
CALC THI: 3
-----------

14
-----------
CELL_HEIGHT: 47
DESCENT: -11
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 1.2636719
CALC POS: -10
CALC THI: 3
-----------

15
-----------
CELL_HEIGHT: 51
DESCENT: -11
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 1.3540039
CALC POS: -10
CALC THI: 3
-----------

16
-----------
CELL_HEIGHT: 54
DESCENT: -12
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 1.4443359
CALC POS: -11
CALC THI: 3
-----------

17
-----------
CELL_HEIGHT: 58
DESCENT: -13
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 1.534668
CALC POS: -12
CALC THI: 3
-----------

18
-----------
CELL_HEIGHT: 61
DESCENT: -14
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 1.625
CALC POS: -12
CALC THI: 4
-----------

19
-----------
CELL_HEIGHT: 64
DESCENT: -14
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 1.7148438
CALC POS: -13
CALC THI: 4
-----------

20
-----------
CELL_HEIGHT: 68
DESCENT: -15
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 1.8051758
CALC POS: -14
CALC THI: 4
-----------

21
-----------
CELL_HEIGHT: 71
DESCENT: -16
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 1.8955078
CALC POS: -14
CALC THI: 4
-----------
```

</details>

chrisduerr commented:
Hmm that's odd. If I calculate the position myself from the metrics you've provided, I get a result that's still within the cell bounds.

Unfortunately I'm also not able to reproduce it on my system which kinda makes this a bit hard to troubleshoot for me. When using consolas myself (size 20) it looks like this:

![](https://u.teknik.io/kypPn.png)

It seems to be well within bounds.

maximbaz commented on src/display.rs:
```diff
@@ -420,3 +481,87 @@ impl Display {
+            let y = ((start.line.0 as f32 + 1.) * metrics.line_height as f32 + metrics.descent
+                - metrics.underline_position
+                - metrics.underline_thickness / 2.)
+                .round() as u32;
```
Just out of curiosity I replaced `.round()` with `.floor()` here and got these results:
- pixel perfect underline on font sizes [8-18, 21]
- underline beyond cell height on font sizes [5-7, 19-20]

This is obviously not a solution, but it got me wondering: `metrics.*` properties are all rounded in `font/src/ft/mod.rs`, here we have some calculations based on already rounded values and then round again, can't this introduce some increasing calculation error? What if `metrics.*` were not rounded in `font/src/ft/mod.rs`, but left as `f32`, and here we had rounding only once, in the end of all calculations?

maximbaz replied:
In case there are multiple Consolas fonts with different metrics, I use [this one](https://aur.archlinux.org/packages/ttf-consolas-with-powerline/) - try, maybe you can reproduce with it 🙂

chrisduerr replied:
I think rounding multiple times should be fine because these metrics should never be float values. The underline position for example shouldn't be `11.2` pixels below the baseline because the concept of `0.2` pixels doesn't exist. And after we make use of the metrics we have to round again because we need `u32`s, so it's either floor, ceil or round to get valid integers.

I think that approach is correct, however if someone else knows more, please correct me.

Reproducing your issue will probably be hard for me because of potential DPI differences. I've used the `consolas-fonts` package to test though, so they might be different.

The question is just if my "algorithm" is correct really. I can't see anything wrong with it and it should be fine with the values you've supplied. If the algorithm is correct, but the font still specifies an offset that is out of bounds, I think alacritty should handle it.

I'm gonna go ahead and implement a bounds check which will make sure that whenever the underline is out of bounds, the underline will be moved up so it's exactly within bounds (using the thickness provided with the font). This should fix the issue and also help with potentially broken font metrics.

However if my approach to this calculation is wrong and maybe rounding shouldn't take place in the metrics, we should still fix this instead of just relying on a bounds check to catch off-by-one errors.

maximbaz replied:
I added this to your patch:

```
println!("CALC POS F32: {}", f32::from(face.ft_face.underline_position()) * x_scale / 64.);
println!("CALC THI F32: {}", f32::from(face.ft_face.underline_thickness()) * x_scale / 64.);
```

Let's look at a random output:

```
-----------
CELL_HEIGHT: 44
DESCENT: -10
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 1.1733398
CALC POS F32: -8.836716
CALC THI F32: 2.6400146
CALC POS: -9
CALC THI: 3
-----------
```

What I was saying is that because F32 values can have any fractional part, the following hypothetical example can happen:

```
CALC_POS_F32 = 7.49
CALC_THI_F32 = 2.49

(CALC_POS_F32.round() + CALC_THI_F32.round() / 2.0).round() == (7 + 2 / 2).round() == 8
(CALC_POS_F32 + CALC_THI_F32 / 2.0).round() == (8.735).round() == 9
```

Let me know if this makes sense, I'll try the new bounds check in meantime.

chrisduerr replied:
Yeah I understand what you mean. But in my opinion it doesn't make any sense to have a fractional underline position/thickness. It should always be an integer offset.

maximbaz replied:
I don't see any problem with your algorithm either, and just out of curiosity I verified that simply removing `.round()` when calculating `CALC POS` and `CALC THI` doesn't solve my particular problem, so I agree, let's keep the code as it is. Bounds check on the other hand is solving the problem well 🙂

maximbaz commented on src/display.rs:
```diff
@@ -420,3 +481,87 @@ impl Display {
+
+    let (y, height) = match flag {
+        cell::Flags::UNDERLINE => {
+            // Get the baseline positon and offset it down by (-) underline position
```
positon -> position

chrisduerr replied:
Thanks :)

chrisduerr commented:
I've added the bounds check, would be interested to see if that fixes it for you @maximbaz. I'd still love to know why it's rendered outside in the first place.

maximbaz commented:
With the last commit the underline is never exceeding the cell and thus always matches the box cursor perfectly 👍 

By the way, my dpi value is 210, so you can also try `xrandr --dpi 210` and then starting alacritty to see if you can reproduce what I was showing 🙂

chrisduerr commented:
I've tried to reproduce it with `210` DPI but that didn't help either. Not sure what could be going on but would you mind sharing the `y` position of the calculated underline for a size that's broken? Maybe 20?

It's calculated in line 556 of `src/display.rs` (you'll have to disable the `if` check).

maximbaz commented:
For font size 20:

```
-----------
CELL_HEIGHT: 68
DESCENT: -15
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 1.8051758
CALC POS F32: -13.59523
CALC THI F32: 4.0616455
CALC POS: -14
CALC THI: 4
-----------

===========
Y: 201
UNDERLINE_HEIGHT: 4
===========
```

![image](https://user-images.githubusercontent.com/1177900/36356528-ac24eb4c-14f3-11e8-8033-4d8b0afde379.png)

chrisduerr commented:
I forgot to ask you what the line number is but based on the `Y` I'd guess it's `3`.

That `201` seems like a very odd number to me though. Because if `Y` would really be `3`, the bottom line of the cell in line 3 should be at `68 * 3 = 204`. Now if we take the calculated underline thickness, divide it by two and add it to the reported `y = 201`, we're at `203`. Instead of being one pixel below, this should be one pixel above.

If the line number is only `2`, the bottom of the cell would be `136`, so the underline would be off by a ton.

These numbers seem a bit odd to me.

maximbaz commented:
The line number is `3`. Just following the code, the math is:

```
3 (line number) * 68 (line_height) + (-15) (descent) - (-14) (underline position) - (4 / 2) (underline thickness)

3 * 68 - 15 + 14 - 2 = 201
```

chrisduerr commented:
Yeah so that doesn't make any sense at all then. Clearly if the line number is `3` and the line height is `68`, then `201` should be accurate and fit well within the cell.

maximbaz commented:
> If you have another line below one with underline it actually does intersect with it, right? It's not like the block cursor is just not big enough?

I believe yes, but it's hard to tell, which symbol will fill the entire cell besides block cursor?

![image](https://user-images.githubusercontent.com/1177900/36357456-84ecd0ac-14fe-11e8-9ef3-3c7e2d064bfb.png)

And with the latest commit in this PR:

![image](https://user-images.githubusercontent.com/1177900/36357357-362087ee-14fd-11e8-9cce-fc9f735c02a3.png)

It is curious to investigate, but I feel we should leave this with the approach taken in the last commit 😄

chrisduerr commented:
The best character to render for filling the whole block is the box cursor. The unicode character for it is defined here:
https://github.com/jwilm/alacritty/blob/master/font/src/lib.rs#L74

chrisduerr commented:
Looking at it closely, it definitely seems to intersect:
![](https://u.teknik.io/jBcgd.png)

The problem is not the case you have, but if it's not an edge case. It would be possible that a single pixel underline is rendered one pixel too far down for example. Assuming that it is an issue with the way I calculate things.

However I'm fine with letting it go the way it is right now. The math seems accurate from my point of view and I don't think there's gonna be any configuration that would look bad. And if there is one that looks really out of place, we can just act upon it when we find it.

maximbaz commented:
`BOX_CURSOR_CHAR` is intersecting just as well. I agree, with bounds check it looks very good and let's leave it as it is for now 🙂

chrisduerr commented:
Unfortunately I'm getting GL errors on my macos VM right now. So I'm not able to implement and test this on macos.

petobens commented:
Hi! Not sure if the strikethrough is already implemented on mac but it doesn't seem to be working:

<img width="2880" alt="screen shot 2018-03-10 at 11 12 38" src="https://user-images.githubusercontent.com/2583971/37243178-2d1dce02-2454-11e8-8460-b55bc986e8b5.png">

Again, I might've understood wrongly your last commit (I believe it adds strikethrough support on mac) so if that's the case sorry about the noise. Thanks for the beautiful work you are all doing with Alacritty :)

chrisduerr commented:
I can not test macos because I am currently not able to get any macos VM to work. If the metrics are calculated properly it will also work on macos, however I am not able to test the metrics so I can't implement the macos part of this.

petobens commented:
@chrisduerr thanks for the info!

jwilm commented:
@chrisduerr would you mind committing the script that generates this?

Demo:
![](https://u.teknik.io/Nwzo3.png)

chrisduerr commented:
```bash
echo -e "RESET \e[4;32m UNDERyLINE \e[0m RESET \e[9;31m STRIKEyTHROUGH \e[0m RESET \e[9m\e[4;35m BOTH till the endy of the line \U10a3e4"
```

chrisduerr commented:
MacOS still needs to get the proper underline and strikethrough metrics. Until then I would recommend against merging this. It shoud be simple to implement, it's just a matter of time to find someone who can test on macos.

malramsay64 commented:
This is what it looks like on macOS with the font Iosevka, other fonts I have tried have similar behavior.

<img width="559" alt="screen shot 2018-03-22 at 07 18 34" src="https://user-images.githubusercontent.com/10113564/37736220-750f95e8-2da4-11e8-9f42-ab74678c6366.png">

I calculated the initial position metrics as below.
```
﻿-----------
CELL_HEIGHT: 37
DESCENT: 7
UNDERLINE POS: -1.0527344
UNDERLINE THI: 2.1191406
STRIKETHROUGH POS: 25.5
STRIKETHROUGH THI: 2.1191406
CALC POS: -1.0527344
CALC THI: 2.1191406
-----------
```
By modifying the position metrics moving the underline down by `- descent` and changing the strikeout from `+ descent` to `- descent` the position of the underline and strikethrough appear in reasonable positions over a range of font sizes ([patch here](https://gist.github.com/malramsay64/01ae2bdd3209aee55c7e75383d9cfa81) )

![screen shot 2018-03-22 at 08 00 52](https://user-images.githubusercontent.com/10113564/37737192-2feb7a60-2da7-11e8-921f-3effab311e64.png)

With position metrics
```
-----------
CELL_HEIGHT: 37
DESCENT: 7
UNDERLINE POS: -8.052734
UNDERLINE THI: 2.1191406
STRIKETHROUGH POS: 11.5
STRIKETHROUGH THI: 2.1191406
CALC POS: -8.052734
CALC THI: 2.1191406
-----------
```
Additionally with some font sizes there appears to be issues with the x positioning of the underline and strikethough.  With the strikethrough going into the reset and starting partway through the B in the below image.

![screen shot 2018-03-22 at 07 59 25](https://user-images.githubusercontent.com/10113564/37737497-f5ffe89e-2da7-11e8-98a7-cb91944c5df0.png)

chrisduerr commented:
Thanks a lot for looking into this! No clue why the X position is wrong, I'll have to look into it again when I have the time. Do you use any glyph offsets?

And your patch seems to work well enough, can you test it with massive font sizes too? I've noticed on Linux that my original approach used unscaled positions, so please make sure that the width of the underline gets bigger with bigger fonts and the position stays reasonable.

Also feel free to send a PR to my branch if you want to, it would be great if someone could add the final tweaks to macos so we can land this. :)

malramsay64 commented:
At font size 72 on a retina display

<img width="1403" alt="screen shot 2018-03-22 at 11 29 17" src="https://user-images.githubusercontent.com/10113564/37744633-4e0f9c34-2dc4-11e8-8790-8e67c53d4914.png">

I got mixed up with the testing I did above. The issues with the X positioning are both font and font size dependent. Iosevka, and Fira Code work fine across the entire range I tested while Monaco has the positioning issue.

PR Sent!

chrisduerr commented:
It seems like `font.offset`, `font.glyph_offset` and `window.padding` all aren't handled properly and can lead to underline/strikethrough not matching the glyph positions.

Either way this should not get merged until #963 is accepted.

mpcsh commented:
Hey, what's the status on this? Contemplating switching to alacritty but the lack of underline is bugging me. I'm happy to contribute/remote-pair if you're interested @chrisduerr.

chrisduerr commented:
@mpcsh Since this is a bigger change it will probably have to wait until the scrollback branch is merged before I can rebase and merge this.

petobens commented:
Is it possible to add undercurl support? (tmux will eventually support it as per https://github.com/tmux/tmux/issues/1492)

BTW: since the scrollback branch is merged is this PR going to be rebased/prioritize? (it's a really nice feature :))

chrisduerr commented:
@petobens That would be significantly more work than a simple underline. And I don't think many terminal emulator support that yet. I also don't think it's standardized in any way.

So this is definitely not within the scope of the PR and if you are interested in it you should open a separate issue.

alexanderjeurissen commented:
@chrisduerr I was looking into underline support remembering seeing a PR for it a while back (which is this PR 😄) In July you mentioned wanting to await merge-in of the  `scroll-back` branch, is this branch dependent on some changes on that branch ? or what is holding us back from rebasing this and merging it in ?

chrisduerr commented:
The primary priority right now is #1403, which depends on an update to https://github.com/tomaka/glutin.

After that I can keep working my way through the open PRs. I've finished #1188 just today and #1190 is another one which will definitely be merged before this.

Once there are no newer PRs left open, I'll get to this.

nakulj commented:
Would it be possible to have underlines skip descenders? Chrome started doing this with URL underlines a little while ago and now I don't want to go back.

![skip_descenders](https://user-images.githubusercontent.com/3213360/48370030-f836d680-e66c-11e8-8636-cbd1593371cc.png)

Perhaps this could be a configurable, but it would be nice to have the option.

chrisduerr commented:
Based on initial estimation, I'd say that's out of scope. It seems way too complicated for what it's worth. If Alacritty ever starts using some kind of font library which helps with this, maybe. But it's not looking like this is going to happen anytime soon (or ever).

chrisduerr commented:
With a bit of work I was able to rebase this branch back on master and make it work again. It seems to be performing fairly well in the minimal tests I've done, so I don't expect any regressions from the state it was in when it was put on hold.

font.offset, font.glyph_offset and window.padding

However, there are still a few things to resolve:
 - [ ] Account for font and glyph offsets
 - [ ] Account for window padding
 - [ ] Test (and potentially fix) DPI
 - [ ] Refactor code and extract it into separate files

dm1try commented:
`osx(10.14.2)`
`Source Code Pro 13.0`
DPI/padding/offsets are default

terminal & iTerm2 & alacritty:
<img width="724" alt="screen shot 2018-12-20 at 11 39 42" src="https://user-images.githubusercontent.com/486807/50273716-70eb3880-044c-11e9-94d1-807d2647baae.png">

chrisduerr commented:
@dm1try That's not with an offset of 0, a glyph offset of 0, a padding of 0 and 1.0 DPI though, right?

dm1try commented:
> That's not with an offset of 0, a glyph offset of 0, a padding of 0

I used default settings(offsets were commented) and I've tried with explicit set to `0`
```
# Configuration for Alacritty, the GPU enhanced terminal emulator
#env:
  # TERM env customization. Default is xterm-256color
  # Note: the default TERM value `xterm-256color` does not
  # specify all features alacritty supports. This does pose
  # a few issues with programs relying on terminfo and the
  # `tput` command
  # TERM: alacritty

window:
  # Window dimensions in character columns and lines
  # (changes require restart)
  #  dimensions:
  #  columns: 0
  #  lines: 0

  # Adds this many blank pixels of padding around the window
  # Units are physical pixels; this is not DPI aware.
  # (change requires restart)
  padding:
    x: 0
    y: 0

  # Window decorations
  # Setting this to false will result in window without borders and title bar.
  decorations: full
  fullscreen: true
  start_maximized: false

# Display tabs using this many cells (changes require restart)
tabspaces: 8

# When true, bold text is drawn using the bright variant of colors.
draw_bold_text_with_bright_colors: true

# Font configuration (changes require restart)
font:
  # The normal (roman) font face to use.
  normal:
    family: "Source Code Pro" # should be "Source Code Pro" or something on macOS.
    # Style can be specified to pick a specific face.

  # The bold font face
  bold:
    family: "Source Code Pro" # should be "Source Code Pro" or something on macOS.
    # Style can be specified to pick a specific face.
    # style: Bold

  # The italic font face
  italic:
    family: "Source Code Pro" # should be "Source Code Pro" or something on macOS.
    # Style can be specified to pick a specific face.
    # style: Italic
  offset:
    x: 0
    y: 0

  # Glyph offset determines the locations of the glyphs within their cells with
  # the default being at the bottom. Increasing `x` moves the glyph to the right,
  # increasing `y` moves the glyph upwards.
  glyph_offset:
    x: 0
    y: 0

  # Point size of the font
  size: 13.0
  # Offset is the extra space around each character. offset.y can be thought of
  # as modifying the linespacing, and offset.x as modifying the letter spacing.
  # OS X only: use thin stroke font rendering. Thin strokes are suitable
  # for retina displays, but for non-retina you probably want this set to
  # false.
  use_thin_strokes: false
...
```

according to DPI, I use default to device scaling. though DPR is 2 on retina display:
```
reated log file at "/var/folders/8t/0phq2zg12js4d9ycstdlbfkr0000gn/T/Alacritty-57922.log"
[2018-12-20 16:46] [INFO] Welcome to Alacritty.
[2018-12-20 16:46] [INFO] Configuration loaded from /Users/dmitrydedov/.alacritty.yml
[2018-12-20 16:46] [INFO] device_pixel_ratio: 2
[2018-12-20 16:46] [INFO] width: 2048, height: 1536
[2018-12-20 16:46] [INFO] Initializing glyph cache
[2018-12-20 16:46] [INFO] Finished initializing glyph cache in 0.058081714
[2018-12-20 16:46] [INFO] width: 1200, height: 792
[2018-12-20 16:46] [INFO] Cell Size: (15 x 33)
[2018-12-20 16:46] [INFO] Padding: (0 x 0)
[2018-12-20 16:46] [INFO] PTY Dimensions: Line(24) x Column(80)
[2018-12-20 16:46] [INFO] Initialisation complete
[2018-12-20 16:46] [INFO] Font size changed: Size(26) [DPR: 2]
[2018-12-20 16:46] [INFO] width: 1200, height: 792
```
on external display the problem is similar:
![screen shot 2018-12-20 at 16 48 23](https://user-images.githubusercontent.com/486807/50288476-17e4ca00-0477-11e9-99b5-543d3074b490.png)

dm1try commented:
btw, unrelated to this issue, but as you can see on [the screenshot](https://user-images.githubusercontent.com/486807/50273716-70eb3880-044c-11e9-94d1-807d2647baae.png) above the default letter spacing in alacritty is different from `terminal/iTerm2`. setting `offset.x` to `1` make them equal, not sure if it's a bug or expected behavior

chrisduerr commented:
As far as I'm concerned, I've addressed all remaining issues except for Windows support.

So in theory it should work on macOS now if the metrics are calculated correctly. Would be appreciated if you could test again @dm1try, hopefully that resolves your problem.

chrisduerr commented:
Also it looks like rusttype provides neither underline, nor strikeout metrics. As a result of that, I'm not sure if it makes sense at this point to support underline/strikeout at all on Windows at this point.

@zacps What do you think about keeping the old hack in place for Window until fontkit has been landed? It seems like fontkit at least offers the underline metrics, maybe we can even upstream strikeout support. Otherwise our custom strikeout which I've had to implement for macOS would do just fine, the important part is mainly the underline.

dm1try commented:
@chrisduerr thanks, it works now!
<img width="650" alt="screen shot 2018-12-21 at 09 46 00" src="https://user-images.githubusercontent.com/486807/50328735-59bc5180-0505-11e9-8049-f6f209f1f0c4.png">

(checked on both displays)

zacps commented:
Fontkit for Windows is basically done, we could ship it for Windows while I work on Linux fixes. I guess it depends how keen you are to get this through? I'd like to say fontkit would be done in a couple of weeks but unfortunately life doesn't work like that.

chrisduerr commented:
@zacps I'd like to get this merged, then have conPTY as the next significant PR and work my way torwards fontkit. With some fixes to URL launching and similar stuff in between.

I'm totally fine with fixing up fontkit myself on linux, that shouldn't be any problem. I can help out at least a little bit to speed things up. My idea was to get this PR into 0.2.5 and potentially release 0.2.6 with fontkit (at least as an option on macOS/linux, probably the default on Windows).

So for me it doesn't make much sense to try and spin up some mediocre underline/strikeout support for Windows now, when the next release is going to have proper support for it anyways. Just so we don't have to do things twice. The old and hacky way to do URLs can still stay in place for Windows, that shouldn't be a problem, so there's not gonna be any regression.

chrisduerr commented:
So I started writing the code for using the old fallback behavior just for Windows and then I realized that I'm just making life hard on myself. Coming up with some okay-ish metrics for underline and strikeout really isn't difficult, so the much, much easier solution was just to implement these metrics. No need to make it unnecessarily hard and ugly.

Would appreciate some testing though @zacps, since I don't have access to Windows right now. There's no scientific background for the way I made these metrics up, other than them looking decent on linux. So let me know if that works out for you too!

cole-h commented:
I copied the echo command that @dm1try used above to test. Strikethrough doesn't seem to be working, but underline is. Adding a reset escape code `\e[0m` fixes the underline issue present in the image -- I left as is in case this is unexpected. Here's a screenshot of the output in both Powershell and Bash under Ubuntu in WSL, respectively:

![alacritty_2018-12-21_18-55-52](https://user-images.githubusercontent.com/28582702/50369916-0d4b3300-0552-11e9-9db1-3fe8a09338d2.png)

Let me know if there's any other information or testing I can provide.

cole-h commented:
Patch applied, added reset escapes for cleaner screenshot: no underlines on the `STRIKEyTHROUGH`.

![alacritty_2018-12-21_21-48-50](https://user-images.githubusercontent.com/28582702/50371171-527b5f00-056a-11e9-98cf-af8311cb5a80.png)

Let me know if there is anything else!

chrisduerr commented:
Thanks, so I think this is good to do then. I'll do a final code review to make sure everything's clean and working properly and then it'll hopefully be merged.

Thanks a ton for the help again @cole-h!

FichteFoll commented:
Thanks for implementing and updating.

However, when testing this, I have to say the underline position is a bit unfortunate with terminus.

The upper terminal shows an older master version (cf9d94e) with a patch I applied as per https://github.com/jwilm/alacritty/issues/31 and the lower terminal is from current master unpatched:

![2018-12-24_13-38-32](https://user-images.githubusercontent.com/931051/50399696-811f5280-0781-11e9-86b5-536866e9c0e1.png)

[Here's my config](https://github.com/FichteFoll/dotfiles/blob/d17b9f4989421e02eaa0a85d20771762ff58c222/alacritty/.config/alacritty/alacritty.yml#L84).

chrisduerr commented:
@FichteFoll I believe you're absolutely correct and there is actually a bug in the computation of the line metrics. I've created a PR [here](https://github.com/jwilm/alacritty/pull/1921), please let me know if that fixes it.

Also feel free to just open a new issue in the future if you're suspecting that there might be a bug. Feedback is always appreciated!

FichteFoll commented:
Wasn't sure about it being a bug or a problem on my part, so I presumed commenting on the PR that introduced the change was the most appropriate.
